clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -183,22 +183,33 @@ public RecordAppendResult append(TopicPartition tp,
             ByteBuffer buffer = free.allocate(size, maxTimeToBlock);
             synchronized (dq) {
                 // Need to check if producer is closed again after grabbing the dequeue lock.
-                if (closed)
+                if (closed) {
+                    free.deallocate(buffer);
                     throw new IllegalStateException("Cannot send after the producer is closed.");
+                }
+
+                RecordAppendResult appendResult = null;
+                try {
+                    appendResult = tryAppend(timestamp, key, value, callback, dq);
+                } catch (Exception e) {
```

Comment:
Right, even `Utils.notNull` throws NPE. So, it is better to have try catch for the whole block assuming there will be no Exception from `free.deallocate(buffer)` from `appendResult!=null` condition instead of having at 3 different places while catching Exception.

```java
    try {
        RecordAppendResult appendResult = null;
        appendResult = tryAppend(timestamp, key, value, callback, dq);

        if (appendResult != null) {
            // Somebody else found us a batch, return the one we waited for! Hopefully this doesn't happen often...
            free.deallocate(buffer);
        } else {
            MemoryRecordsBuilder recordsBuilder = MemoryRecords.builder(buffer, compression, TimestampType.CREATE_TIME, this.batchSize);
            RecordBatch batch = new RecordBatch(tp, recordsBuilder, time.milliseconds());
            FutureRecordMetadata future = Utils.notNull(batch.tryAppend(timestamp, key, value, callback, time.milliseconds()));

            dq.addLast(batch);
            incomplete.add(batch);
            appendResult = new RecordAppendResult(future, dq.size() > 1 || batch.isFull(), true);
        }

        return appendResult;
    } catch (Exception e) {
        free.deallocate(buffer);
        throw e;
    }
```